### DISCUSSION TITLE
SM2 signature verification failed using the x, y coordinates.

### DISCUSSION THREAD

**Discussion by Nichanghao-xiaohao** [openssl/openssl#23629] [Q&A] [answered]
In OpenSSL 3.+ version, when deriving the SM2 public key from x, y coordinates to perform signature verification, a segmentation fault occurs at EVP_DigestVerifyUpdate. The following is an example of the code:
`
void print_ossl_params(const OSSL_PARAM *params) {
    if (params == NULL) {
        printf("No parameters to print.\n");
        return;
    }

    for (const OSSL_PARAM *p = params; p->key != NULL; ++p) {
        printf("Parameter: %s\n", p->key);

        switch (p->data_type) {
            case OSSL_PARAM_INTEGER:
            case OSSL_PARAM_UNSIGNED_INTEGER: {
                int num;
                if (OSSL_PARAM_get_int(p, &num)) {
                    printf("  Type: Integer\n");
                    printf("  Value: %d\n", num);
                }
                break;
            }
            case OSSL_PARAM_UTF8_STRING: {
                char str[1024];
                char *ptr_to_buffer = str;
                if (OSSL_PARAM_get_utf8_string(p, &ptr_to_buffer, sizeof(str))) {
                    printf("  Type: String\n");
                    printf("  Value: %s\n", str);
                }
                break;
            }

            default:
                printf("  Type: Unknown\n");
                break;
        }
    }
}

int main() {
    std::cout << "OpenSSL version (compiled): " << OPENSSL_VERSION_TEXT << std::endl;
    std::cout << "OpenSSL version (runtime): "  << OpenSSL_version(OPENSSL_VERSION) << std::endl;

    BN_CTX* ctx = BN_CTX_new();
    const char* hex_x = "529387D6912A3394521BA49A9EC4E2E713165CBDB241A0286F9D80EFFA207637";
    const char* hex_y = "ED40BF2954B37341254F6EFA8E415D63DCD366A9307E1BC7032D2F7B42CC3E6E";
    EVP_PKEY_CTX* pctx = EVP_PKEY_CTX_new_from_name(NULL, "SM2", NULL);

    BIGNUM* bn_x = BN_new();
    BIGNUM* bn_y = BN_new();
    BN_hex2bn(&bn_x, hex_x);
    BN_hex2bn(&bn_y, hex_y);

    OSSL_PARAM_BLD* param_bld = OSSL_PARAM_BLD_new();
    OSSL_PARAM_BLD_push_utf8_string(param_bld, OSSL_PKEY_PARAM_GROUP_NAME, "SM2", 0);
    OSSL_PARAM_BLD_push_BN(param_bld, OSSL_PKEY_PARAM_EC_PUB_X, bn_x);
    OSSL_PARAM_BLD_push_BN(param_bld, OSSL_PKEY_PARAM_EC_PUB_Y, bn_y);
    
    OSSL_PARAM* params = OSSL_PARAM_BLD_to_param(param_bld);
    print_ossl_params(params);
    EVP_PKEY *pkey = NULL;
    if (!EVP_PKEY_fromdata_init(pctx))
    {
        std::cerr << "Unable EVP_PKEY_fromdata_init" << std::endl;
    }
    if (!EVP_PKEY_fromdata(pctx, &pkey, EVP_PKEY_PUBLIC_KEY, params)) {
        std::cerr << "Unable to set public key." << std::endl;
        ERR_print_errors_fp(stderr);
        EVP_PKEY_CTX_free(pctx);
        EVP_PKEY_free(pkey);
        OSSL_PARAM_free(params);
        OSSL_PARAM_BLD_free(param_bld);
        BN_free(bn_x);
        BN_free(bn_y);
        return 1;
    }
    EVP_MD_CTX *md_ctx = EVP_MD_CTX_new();
    EVP_PKEY_CTX* pkeyCtx = EVP_PKEY_CTX_new(pkey, NULL);
    EVP_PKEY_CTX_set1_id(pkeyCtx, "1234567812345678", 16);
    EVP_MD_CTX_set_pkey_ctx(md_ctx, pctx);
    EVP_DigestVerifyInit(md_ctx, &pkeyCtx, EVP_sm3(), NULL, pkey);

    std::cerr << "md_ctx:" << md_ctx << std::endl;
    if(EVP_DigestVerifyUpdate(md_ctx, "1234567812345678", 16))
    {
        ERR_print_errors_fp(stderr);
    }

    // EVP_PKEY_CTX_free(pctx);
    // EVP_PKEY_free(pkey);
    // OSSL_PARAM_free(params);
    // OSSL_PARAM_BLD_free(param_bld);

    return 0;
}

`
**OutPut:**
OpenSSL version (compiled): OpenSSL 3.2.0 23 Nov 2023
OpenSSL version (runtime): OpenSSL 3.2.0 23 Nov 2023
Parameter: group
  Type: String
  Value: SM2
Parameter: qx
Parameter: qy
md_ctx:0x13ff11070
**zsh: segmentation fault.**
![6C1D4AB8-D77F-4DCD-B4C0-7A6E3093AE0D](https://github.com/openssl/openssl/assets/55940146/31cd8e5c-0942-4053-84bd-35c2bd376b9b)

**Comment by t8m** [comment, marked as answer]
OSSL_PKEY_PARAM_EC_PUB_X and OSSL_PKEY_PARAM_EC_PUB_Y parameters are not importable (i.e., they are read-only). That means the imported pkey contains only the SM2 parameters but not a public key. You have to use the encoded form of the public key to import via the OSSL_PKEY_PARAM_PUB_KEY parameter { 0x04, x value bytes, y value bytes }.

Also to avoid segfaults in such cases, you have to check results of the previous operations for errors.

**Reply by Nichanghao-xiaohao** [reply]
Thank you for your response; it helped me solve an issue that had been troubling me, and now I am able to correctly generate the public key. The modified code is as follows:
```
    EVP_PKEY *pkey = NULL;
    OSSL_PARAM_BLD *param_bld = NULL;
    OSSL_PARAM *params = NULL;
    BIGNUM *x = NULL, *y = NULL;
    unsigned char public_key[65];
    public_key[0] = 0x04;

    if (!BN_hex2bn(&x, "529387D6912A3394521BA49A9EC4E2E713165CBDB241A0286F9D80E"
                       "FFA207637") ||
        BN_hex2bn(&y, "ED40BF2954B37341254F6EFA8E415D63DCD366A9307E1BC7032D2F7B"
                      "42CC3E6E")) {
        std::cerr << "Error converting hex to BIGNUM.\n";
    }

    BN_bn2binpad(x, public_key + 1, 32);
    BN_bn2binpad(y, public_key + 1 + 32, 32);
    EVP_PKEY_CTX *pctx = EVP_PKEY_CTX_new_from_name(NULL, "SM2", NULL);
    if (!pctx)
    {
        std::cerr << "Error creating EVP_PKEY_CTX.\n";
    }
    param_bld = OSSL_PARAM_BLD_new();
    if (!param_bld) 
    {
        std::cerr << "Error creating OSSL_PARAM_BLD.\n";
    }
    if (!OSSL_PARAM_BLD_push_utf8_string(param_bld, OSSL_PKEY_PARAM_GROUP_NAME, "SM2", 0))
    {
        std::cerr << "Error setting group name." << std::endl;
    }
    
    if (!OSSL_PARAM_BLD_push_octet_string(param_bld, OSSL_PKEY_PARAM_PUB_KEY, public_key, 65))
    {
        std::cerr << "Error pushing public key to param builder.\n";
    }
    
    params = OSSL_PARAM_BLD_to_param(param_bld);
    if (EVP_PKEY_fromdata_init(pctx) <= 0 ||
        EVP_PKEY_fromdata(pctx, &pkey, EVP_PKEY_PUBLIC_KEY, params) <= 0) {
        ERR_print_errors_fp(stderr);
        std::cerr << "Error: EVP_PKEY_fromdata failed\n";
    }

    FILE *pem_file = fopen("sm2_pubkey.pem", "w");
    if (pem_file == NULL) {
        std::cerr << "Error: Cannot open file\n";
    }
    PEM_write_PUBKEY(pem_file, pkey);
    EVP_MD_CTX *md_ctx = EVP_MD_CTX_new();
    EVP_PKEY_CTX* pkeyCtx = EVP_PKEY_CTX_new(pkey, NULL);
    EVP_DigestVerifyInit(md_ctx, &pkeyCtx, EVP_sm3(), NULL, pkey);
    if(EVP_DigestVerifyUpdate(md_ctx, "1234567812345678", 16))
    {
        ERR_print_errors_fp(stderr);
    }
```

**Reply by Nichanghao-xiaohao** [reply]
Using the code above, the output of EVP_PKEY_base_id(pkey) is 0, and when using EVP_DigestVerifyFinal, there is an error: ossl_sm2_internal_verify:reason(104):/usr/src/debug/openssl/3.1.0-r0/crypto/sm2/sm2_sign.c:488:.

**Reply by t8m** [reply]
`EVP_PKEY_base_id(pkey) == 0` for keys coming from providers. That's OK.

The error on EVP_DigestVerifyFinal() indicates that there is some problem with the data of the signature. That it does not have proper encoding.

**Reply by Nichanghao-xiaohao** [reply]
By using solely the EVP_DigestVerify instead of the three interfaces EVP_DigestVerifyInit, EVP_DigestVerifyUpdate, and EVP_DigestVerifyFinal, signature verification succeeds. What could be the reason for this?

**Reply by Nichanghao-xiaohao** [reply]
int r1 = EVP_PKEY_public_check(pkeyCtx);
        std::cerr << "r1: " << r1 << std::endl;
        if (EVP_PKEY_verify_init(pkeyCtx) <= 0)
        {
            ERR_print_errors_fp(stderr);
            std::cerr << "Verify Init Error!!!" << std::endl;
            return -1;
        }
        EVP_PKEY_CTX_set_signature_md(pkeyCtx, EVP_sm3());
        const unsigned char* r_ptr = sigMsg;
        const unsigned char* s_ptr = sigMsg + 32;
        ECDSA_SIG* sm2_sig = ECDSA_SIG_new();
        BIGNUM* r = BN_bin2bn(r_ptr, 32, NULL);
        BIGNUM* s = BN_bin2bn(s_ptr, 32, NULL);
        BN_print_fp(stdout, r);
        std::cerr << std::endl;
        BN_print_fp(stdout, s);
        std::cerr << std::endl;
        ECDSA_SIG_set0(sm2_sig, r, s);
        unsigned char* der = NULL;
        size_t sig_len = i2d_ECDSA_SIG(sm2_sig, &der);
        int ret = EVP_PKEY_verify(pkeyCtx, der, sig_len,
                                  reinterpret_cast<const unsigned char*>(msg),
                                  msgLen);
        std::cerr << "ret: " << ret << std::endl;  why does EVP_PKEY_verify return 0 when r1, r, and s are all correct in the aforementioned code?

**Reply by t8m** [reply]
EVP_PKEY_verify() expects the hash of the message and not the message itself.

**Reply by Nichanghao-xiaohao** [reply]
unsigned char hashMsg[32]; 
        unsigned int hashMsgLen = 0;
        EVP_Digest(reinterpret_cast<const unsigned char*>(msg), msgLen,
                       hashMsg, &hashMsgLen, EVP_sm3(), NULL);
        int ret = EVP_PKEY_verify(pkeyCtx, der, sig_len, hashMsg, 32);    Even after modifying it as shown above, the value of ret remains 0.

**Reply by Nichanghao-xiaohao** [reply]
**The following example outputs 1.**
` uint8_t pubkey[] = {
      0x30, 0x5A, 0x30, 0x14, 0x06, 0x08, 0x2A, 0x81, 0x1C, 0xCF, 0x55, 0x01,
      0x82, 0x2D, 0x06, 0x08, 0x2A, 0x81, 0x1C, 0xCF, 0x55, 0x01, 0x82, 0x2D,
      0x03, 0x42, 0x00, 0x04, 0x77, 0x58, 0x36, 0x91, 0x2E, 0x19, 0x5E, 0x0A,
      0xD0, 0x5A, 0xA1, 0x17, 0x51, 0x5E, 0x1F, 0x18, 0xB9, 0xF8, 0xB3, 0xCB,
      0xD3, 0xAE, 0xE6, 0x73, 0x9D, 0x6B, 0xAC, 0xE9, 0x08, 0xAB, 0x47, 0x2C,
      0xDF, 0x2C, 0x6F, 0x3D, 0x85, 0x04, 0x12, 0x13, 0xB0, 0x9F, 0x2D, 0x03,
      0xCE, 0xEE, 0x86, 0xBB, 0x12, 0xEB, 0x7F, 0x08, 0x48, 0x96, 0xB1, 0xDC,
      0xF8, 0xD0, 0xD8, 0xDC, 0xAA, 0x42, 0x9B, 0x20,
  };

  uint8_t *pubkeyPointer = pubkey;
  EVP_PKEY *evpPkey = d2i_PUBKEY(NULL, &pubkeyPointer, sizeof(pubkey));

  EVP_PKEY_CTX *ctx = EVP_PKEY_CTX_new(evpPkey, NULL);
  EVP_PKEY_verify_init(ctx);
  EVP_PKEY_CTX_set_signature_md(ctx, EVP_sm3());

  uint8_t sig[] = {
      0x30, 0x46, 0x02, 0x21, 0x00, 0x9E, 0x2C, 0xA7, 0xF2, 0x4E, 0x80, 0x54,
      0x9D, 0xE7, 0x9D, 0x1B, 0x97, 0x38, 0x0E, 0x88, 0xEA, 0x09, 0xE1, 0x46,
      0x50, 0x27, 0x88, 0xD5, 0x1C, 0xB9, 0xCC, 0xED, 0x7D, 0x43, 0x32, 0x50,
      0x3C, 0x02, 0x21, 0x00, 0x81, 0x15, 0x61, 0xD7, 0x9A, 0xEE, 0x20, 0x3C,
      0x4D, 0x92, 0x41, 0xBF, 0xD5, 0x8B, 0x38, 0x7C, 0x24, 0x91, 0x3A, 0x11,
      0x18, 0x15, 0x58, 0xCD, 0xC1, 0x99, 0x18, 0x28, 0xBF, 0x8A, 0x0B, 0x73,
  };
  uint8_t plainDigest[] = {
      0x15, 0x9C, 0xA3, 0xD6, 0x6A, 0x79, 0x7D, 0xD5, 0xF5, 0x90, 0x5A,
      0x9F, 0x87, 0x19, 0xBD, 0x41, 0x21, 0xE7, 0x26, 0xEA, 0xB8, 0x83,
      0x7F, 0x29, 0x64, 0x09, 0x3C, 0xA5, 0x19, 0xB3, 0x9F, 0x54,
  };

  int ret =
      EVP_PKEY_verify(ctx, sig, sizeof(sig), plainDigest, sizeof(plainDigest));
  std::cerr << "ret: " << ret << std::endl;
  return 0;`

**Reply by Nichanghao-xiaohao** [reply]
**The following example outputs 0.**

`  EVP_PKEY *pkey = NULL;
  OSSL_PARAM_BLD *param_bld = NULL;
  OSSL_PARAM *params = NULL;
  BIGNUM *x = NULL, *y = NULL;
  unsigned char public_key[65];
  public_key[0] = 0x04;

  if (!BN_hex2bn(
          &x,
          "529387D6912A3394521BA49A9EC4E2E713165CBDB241A0286F9D80EFFA207637") ||
      !BN_hex2bn(
          &y,
          "ED40BF2954B37341254F6EFA8E415D63DCD366A9307E1BC7032D2F7B42CC3E6E")) {
    std::cerr << "Error converting hex to BIGNUM.\n";
  }

  BN_bn2binpad(x, public_key + 1, 32);
  BN_bn2binpad(y, public_key + 1 + 32, 32);
  EVP_PKEY_CTX *pctx = EVP_PKEY_CTX_new_from_name(NULL, "SM2", NULL);
  if (!pctx) {
    std::cerr << "Error creating EVP_PKEY_CTX.\n";
  }
  param_bld = OSSL_PARAM_BLD_new();
  if (!param_bld) {
    std::cerr << "Error creating OSSL_PARAM_BLD.\n";
  }
  if (!OSSL_PARAM_BLD_push_utf8_string(param_bld, OSSL_PKEY_PARAM_GROUP_NAME,
                                       "SM2", 0)) {
    std::cerr << "Error setting group name." << std::endl;
  }

  if (!OSSL_PARAM_BLD_push_octet_string(param_bld, OSSL_PKEY_PARAM_PUB_KEY,
                                        public_key, 65)) {
    std::cerr << "Error pushing public key to param builder.\n";
  }

  params = OSSL_PARAM_BLD_to_param(param_bld);
  if (EVP_PKEY_fromdata_init(pctx) <= 0 ||
      EVP_PKEY_fromdata(pctx, &pkey, EVP_PKEY_PUBLIC_KEY, params) <= 0) {
    ERR_print_errors_fp(stderr);
    std::cerr << "Error: EVP_PKEY_fromdata failed\n";
  }
  EVP_PKEY_CTX *ctx = EVP_PKEY_CTX_new(evpPkey, NULL);
  EVP_PKEY_verify_init(ctx);
  EVP_PKEY_CTX_set_signature_md(ctx, EVP_sm3());

  uint8_t sig[] = {
      0x30, 0x46, 0x02, 0x21, 0x00, 0x9E, 0x2C, 0xA7, 0xF2, 0x4E, 0x80, 0x54,
      0x9D, 0xE7, 0x9D, 0x1B, 0x97, 0x38, 0x0E, 0x88, 0xEA, 0x09, 0xE1, 0x46,
      0x50, 0x27, 0x88, 0xD5, 0x1C, 0xB9, 0xCC, 0xED, 0x7D, 0x43, 0x32, 0x50,
      0x3C, 0x02, 0x21, 0x00, 0x81, 0x15, 0x61, 0xD7, 0x9A, 0xEE, 0x20, 0x3C,
      0x4D, 0x92, 0x41, 0xBF, 0xD5, 0x8B, 0x38, 0x7C, 0x24, 0x91, 0x3A, 0x11,
      0x18, 0x15, 0x58, 0xCD, 0xC1, 0x99, 0x18, 0x28, 0xBF, 0x8A, 0x0B, 0x73,
  };
  uint8_t plainDigest[] = {
      0x15, 0x9C, 0xA3, 0xD6, 0x6A, 0x79, 0x7D, 0xD5, 0xF5, 0x90, 0x5A,
      0x9F, 0x87, 0x19, 0xBD, 0x41, 0x21, 0xE7, 0x26, 0xEA, 0xB8, 0x83,
      0x7F, 0x29, 0x64, 0x09, 0x3C, 0xA5, 0x19, 0xB3, 0x9F, 0x54,
  };

  int ret =
      EVP_PKEY_verify(ctx, sig, sizeof(sig), plainDigest, sizeof(plainDigest));
  std::cerr << "ret: " << ret << std::endl;
`

**Reply by Nichanghao-xiaohao** [reply]
What is the difference between the two examples mentioned above?

**Reply by t8m** [reply]
The public key in the pubkey array has completely different X and Y values than what you use in the second example.
